okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/ExchangeFinder.kt

```diff
@@ -355,6 +323,7 @@ class ExchangeFinder(
 
     synchronized(connection) {
       if (connection.routeFailureCount != 0) return null
+      if (!connection.noNewExchanges) return null // This route is still in use.
```

Comment:
This replaces what we did in trackFailure(): rather than removing a duplicated route we never add it.